[testharnessreport.js] Only pay attention to the top-level completion handler

gsnedders opened the pull request:
#### 5273909547b13c75ce88e806b773737baa363c9d
<pre>
[testharnessreport.js] Only pay attention to the top-level completion handler
<a href="https://bugs.webkit.org/show_bug.cgi?id=259409">https://bugs.webkit.org/show_bug.cgi?id=259409</a>
rdar://problem/112683033

Reviewed by Jonathan Bedard.

Previously, we were considering the test as having to run to completion
when any testharness.js completion handler first ran (which was
sometimes that of the frame within the opened window; the nondeterminism
here just adds flakiness to the already bad behaviour).

Instead, only output anything for the top-level completion handler, as
all results should be passed up to it. Note wptrunner with the WebDriver
or Marionette executors only ever pays attention to the top-level
completion handler (as they only pay attention to the top-level frame &amp;
window), thus they don&apos;t have any flakiness like this.

Ideally testharness.js would have an API we could use for this; I&apos;ve
filed an RFC at <a href="https://github.com/web-platform-tests/rfcs/pull/168">https://github.com/web-platform-tests/rfcs/pull/168</a> for
this, but there&apos;s no reason not to do the simple fix ourselves, as at
least for now this is a strict progression. (It would stop being a
strict progression if at some point testharness.js started allowing
fetch_tests_from_window() with a window with noopener.)

It seems likely we have other tests that are marked as flaky because of
it, but alas there&apos;s no easy way to find what tests will have been fixed
by this.

* LayoutTests/TestExpectations:
* LayoutTests/http/tests/security/contentSecurityPolicy/upgrade-insecure-requests/basic-upgrade-cors.https-expected.txt:
* LayoutTests/http/tests/security/contentSecurityPolicy/upgrade-insecure-requests/basic-upgrade-cors.https.html:
* LayoutTests/imported/w3c/web-platform-tests/cookies/partitioned-cookies/partitioned-cookies.tentative.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/source/navigate-child-function-parent-then-fragment-expected.txt:
* LayoutTests/platform/wk2/TestExpectations:
* LayoutTests/resources/testharnessreport.js:

Canonical link: <a href="https://commits.webkit.org/269483@main">https://commits.webkit.org/269483@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/45dd7a805c8e870165c3cde8e623d3c4755ccc75

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22522 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/179 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23605 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24429 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/20851 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/1247 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/23055 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/21836 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22762 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19546 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25281 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/126 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20405 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/26656 "Passed tests") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20649 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24499 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/107 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/17965 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/81 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/20209 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5403 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/128 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/123 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

JonWBedard commented on LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/source/navigate-child-function-parent-then-fragment-expected.txt:
```diff
@@ -1 +1,9 @@
+
+Harness Error (TIMEOUT), message = null
+
+TIMEOUT
```
I'm concerned to see this. How long does this test take to run with this change? It's not necessarily a problem if the test runs quickly, but taking 30 seconds to run a passing test is clearly bad.

gsnedders replied:
Yeah, we should perhaps skip this test. Though it does run into the problem of skipping timing out tests in general: https://bugs.webkit.org/show_bug.cgi?id=248694

And the prior behaviour of an empty expectation file is clearly bogus.

(Note that this does already timeout on wptrunner: https://wpt.fyi/results/html/browsers/browsing-the-web/navigating-across-documents/source/navigate-child-function-parent-then-fragment.html?run_id=5088436001701888&run_id=4800231314554880&run_id=5072135761952768&run_id=5067839888687104)

gsnedders committed:
[testharnessreport.js] Only pay attention to the top-level completion handler
https://bugs.webkit.org/show_bug.cgi?id=259409
rdar://problem/112683033

Reviewed by Jonathan Bedard.

Previously, we were considering the test as having to run to completion
when any testharness.js completion handler first ran (which was
sometimes that of the frame within the opened window; the nondeterminism
here just adds flakiness to the already bad behaviour).

Instead, only output anything for the top-level completion handler, as
all results should be passed up to it. Note wptrunner with the WebDriver
or Marionette executors only ever pays attention to the top-level
completion handler (as they only pay attention to the top-level frame &
window), thus they don't have any flakiness like this.

Ideally testharness.js would have an API we could use for this; I've
filed an RFC at https://github.com/web-platform-tests/rfcs/pull/168 for
this, but there's no reason not to do the simple fix ourselves, as at
least for now this is a strict progression. (It would stop being a
strict progression if at some point testharness.js started allowing
fetch_tests_from_window() with a window with noopener.)

It seems likely we have other tests that are marked as flaky because of
it, but alas there's no easy way to find what tests will have been fixed
by this.

* LayoutTests/TestExpectations:
* LayoutTests/http/tests/security/contentSecurityPolicy/upgrade-insecure-requests/basic-upgrade-cors.https-expected.txt:
* LayoutTests/http/tests/security/contentSecurityPolicy/upgrade-insecure-requests/basic-upgrade-cors.https.html:
* LayoutTests/imported/w3c/web-platform-tests/cookies/partitioned-cookies/partitioned-cookies.tentative.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/source/navigate-child-function-parent-then-fragment-expected.txt:
* LayoutTests/platform/wk2/TestExpectations:
* LayoutTests/resources/testharnessreport.js:

Canonical link: https://commits.webkit.org/269483@main
